src/policy/ancestor_packages.cpp

```diff
@@ -0,0 +1,262 @@
+    return std::make_pair(entry.fee.value(), entry.vsize.value());
+}
+
+std::optional<std::pair<CAmount, int64_t>> AncestorPackage::FilteredAncestorFeeAndVsize(const CTransactionRef& tx) const
+{
+    const auto& entry_it = m_txid_to_entry.find(tx->GetHash());
+    if (entry_it == m_txid_to_entry.end()) return std::nullopt;
+    const auto& entry = entry_it->second;
+    if (entry.m_state == PackageEntry::State::DANGLING || !entry.fee.has_value() || !entry.vsize.has_value()) return std::nullopt;
+
+    CAmount total_fee{0};
+    int64_t total_vsize{0};
+    for (const auto& txid : entry.ancestor_subset) {
+        const auto& anc_entry = m_txid_to_entry.at(txid);
+        // All non-PACKAGE ancestor_subset transactions should have been removed when the
+        // transaction was first skipped.
+        if (Assume(anc_entry.m_state == PackageEntry::State::PACKAGE)) {
+            if (anc_entry.fee.has_value() && anc_entry.vsize.has_value()) {
+                total_fee += anc_entry.fee.value();
+                total_vsize += anc_entry.vsize.value();
+            } else {
+                // If any fee or vsize information is missing, we can't return an accurate result.
+                return std::nullopt;
+            }
+        }
+    }
+
+    return std::make_pair(total_fee, total_vsize);
+}
+
+void AncestorPackage::IsInMempool(const CTransactionRef& transaction)
+{
+    const auto& txid{transaction->GetHash()};
+    if (m_txid_to_entry.count(txid) == 0) return;
+
+    Assume(m_txid_to_entry.at(txid).m_state != PackageEntry::State::DANGLING);
+    m_txid_to_entry.at(txid).m_state = PackageEntry::State::MEMPOOL;
+
+    // Delete this tx from the descendant's ancestor_subset.
+    for (const auto& descendant_txid : m_txid_to_entry.at(transaction->GetHash()).descendant_subset) {
+        m_txid_to_entry.at(descendant_txid).ancestor_subset.erase(txid);
+    }
+}
+void AncestorPackage::IsDanglingWithDescendants(const CTransactionRef& transaction)
+{
+    const auto& txid{transaction->GetHash()};
+    if (m_txid_to_entry.count(txid) == 0) return;
+    m_txid_to_entry.at(txid).m_state = PackageEntry::State::DANGLING;
+
+    // Set all descendants to DANGLING, as they depend on something that is being discarded.
+    // Delete this tx from the descendant's ancestor_subset.
+    for (const auto& descendant_txid : m_txid_to_entry.at(txid).descendant_subset) {
+        Assume(m_txid_to_entry.at(descendant_txid).m_state != PackageEntry::State::MEMPOOL);
+        m_txid_to_entry.at(descendant_txid).m_state = PackageEntry::State::DANGLING;
+        m_txid_to_entry.at(descendant_txid).ancestor_subset.erase(txid);
+    }
+}
+
+void AncestorPackage::AddFeeAndVsize(const uint256& txid, CAmount fee, int64_t vsize)
+{
+    if (m_txid_to_entry.count(txid) == 0) return;
+    m_txid_to_entry.at(txid).fee = fee;
+    m_txid_to_entry.at(txid).vsize = vsize;
```

Comment:
```c++
if (auto e = m_txid_to_entry.find(txid); e != m_txid_to_entry.end()) {
    e->fee = fee;
    e->vsize = vsize;
}
```

(search for txid once rather than three times)

Reply:
Taken